File: unsolved/time-limit/airports.py
#!/bin/env python

#@profile
def find(parents, vert):
    if (parents[vert] != vert):
        parents[vert] = find(parents,parents[vert])
    return parents[vert]

#@profile
def union(parents, ranks, v1, v2):
    r1 = find(parents, v1)
    r2 = find(parents, v2)
    if (r1 != r2):
        rank1 = ranks[r1]
        rank2 = ranks[r2]
        if (rank1 > rank2):
            parents[r2] = r1
        else:
            parents[r1] = r2
            if (rank1 == rank2): ranks[r2] += 1

#@profile
def kruskals(nodes, edges):
    parents = {}
    ranks = {}

    for node in nodes:
        parents[node] = node
        ranks[node] = 0

    mst = set()
    for edge in sorted(edges):
        _,parent,child = edge
        if (find(parents, parent) != find(parents, child)):
            union(parents, ranks, parent, child)
            mst.add(edge)

    return mst,parents


#@profile
#def prims(cnode, links, airport_cost):
#    total_cost = 0
#
#    visited_nodes = {}
#    visited_nodes[cnode] = None
#
#    links_left = links[cnode]
#
#    finished = False
#    while (not finished):
#        finished = True
#        for link in sorted(links_left):
#            cost,node = link
#            if (not node in visited_nodes):
#                total_cost += cost
#
#                links_left.extend(links[node])
#                links_left = list(filter(lambda x: not x[1] in visited_nodes, links_left))
#
#                visited_nodes[node] = None
#                finished = False
#                break
#
#    return visited_nodes,total_cost

#@profile
def main():
    for i in range(int(input())):
        locations, roads, airport_cost = map(int,input().split())
        #print(locations, roads, airport_cost)

        #links = {}
        #for j in range(locations):
        #    links[j+1] = []

        nodes = range(1,locations+1)

        edges = []
        for _ in range(roads):
            parent, child, cost = map(int,input().split())
            if (cost < airport_cost):
                #links[parent].append((cost, child))
                #links[child].append((cost, parent))
                edges.append((cost,parent,child))
                #edges.append((cost,child,parent))

        mst,parents = kruskals(nodes, edges)

        airports = map(lambda x: find(parents, x), parents)
        total_airports = len(set(airports))

        cost = sum(map(lambda x: x[0], mst))
        print(mst)
        total_cost = total_airports*airport_cost + cost

        print("Case #%d: %d %d" % (i+1,total_cost, total_airports))
        input()

main()

File: solved/airports.cpp
#include <iostream>
#include <tuple>
#include <vector>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>

using namespace std;

int find(unordered_map<int,int>& parents,
		 int vert) {
	if (parents[vert] != vert) {
		parents[vert] = find(parents,parents[vert]);
	}
	return parents[vert];
}

void union_set(unordered_map<int,int>& parents,
			   unordered_map<int,int>& ranks,
			   int v1,
			   int v2) {
    int r1 = find(parents, v1);
    int r2 = find(parents, v2);
    if (r1 != r2) {
        int rank1 = ranks[r1];
        int rank2 = ranks[r2];
        if (rank1 > rank2) parents[r2] = r1;
        else {
            parents[r1] = r2;
            if (rank1 == rank2) ranks[r2]++;
		}
	}
}

void kruskals(unordered_map<unsigned int,int>& mst, 
			  unordered_map<int,int>& parents,
			  int locations,
			  vector<tuple<int,int,int,int>>& edges) {

	unordered_map<int, int> ranks;
	for (int i = 1; i < locations+1; i++) {
		parents[i] = i;
		ranks[i] = 0;
	}

	sort(edges.begin(), edges.end());
	for (tuple<int,int,int,int> edge : edges) {
		unsigned int cost = get<0>(edge);
		unsigned int parent = get<1>(edge);
		unsigned int child = get<2>(edge);
		if (find(parents, parent) != find(parents, child)) {
			union_set(parents, ranks, parent, child);
			//unsigned int hash = cost;
			//hash *= 37;
			//hash += parent;
			//hash *= 37;
			//hash += child;
			unsigned int index = get<3>(edge);
			mst[index] = cost;
		}
	}
}

int main() {
	//Should probably switch over to scanf
	//But why is the question time limit so tight
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	int cases;
	cin >> cases;
	for (int i = 0; i < cases; i++) {
		int locations, roads, airport_cost;
		cin >> locations;
		cin >> roads;
		cin >> airport_cost;

		vector<tuple<int,int,int,int>> edges;
		for (int road = 0; road < roads; road++) {
			int parent, child, cost;
			cin >> parent;
			cin >> child;
			cin >> cost;
			if (cost < airport_cost) {
				edges.push_back(make_tuple(cost,parent,child,road));
				//edges.push_back(make_tuple(cost,child,parent));
			}
		}

		unordered_map<unsigned int,int> mst;
		unordered_map<int,int> parents;
		kruskals(mst, parents, locations, edges);

		unordered_set<int> total_airports;
		for (auto edge : parents) {
			total_airports.insert(find(parents,edge.first));
		}

		int total_cost = total_airports.size()*airport_cost;
		for (auto cost : mst) {
			total_cost += cost.second;
		}

		printf("Case #%d: %d %d\n", i+1, total_cost, total_airports.size());
	}
	return 0;
}
